src/lib/storage/buffer/volatile_region.cpp

```diff
@@ -0,0 +1,238 @@
+  // We create one buffer per thread to avoid contention. This buffer is divided into intermediate pages
+  // for each potential NUMA node. For simplicity, we use of the maximum page size to accommodate all page sizes.
+  static thread_local auto intermediate_buffer = []() {
+#if HYRISE_NUMA_SUPPORT
+    static const auto numa_node_count = numa_num_configured_nodes();
+#else
+    static const auto numa_node_count = 1;
+#endif
+    const auto num_bytes = bytes_for_size_type(MAX_PAGE_SIZE_TYPE);
+    auto buffer = std::make_unique_for_overwrite<uint8_t[]>(numa_node_count * num_bytes);
+    for (auto numa_node = uint64_t{0}; numa_node < numa_node_count; ++numa_node) {
+      const auto buffer_offset = buffer.get() + numa_node * num_bytes;
+#if HYRISE_NUMA_SUPPORT
+      numa_tonode_memory(buffer_offset, num_bytes, numa_node);
+#endif
+    }
+    return buffer;
+  }();
+
+  const auto page_ptr = get_page(page_id);
+  const auto target_buffer =
+      intermediate_buffer.get() + static_cast<uint64_t>(target_memory_node) * bytes_for_size_type(MAX_PAGE_SIZE_TYPE);
+  std::memcpy(target_buffer, page_ptr, page_id.byte_count());
+  free(page_id);
+  mbind_to_numa_node(page_id, target_memory_node);
+  std::memcpy(page_ptr, target_buffer, page_id.byte_count());
+}
+
+void VolatileRegion::free(const PageID page_id) {
+  DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+  // Use MADV_FREE_REUSABLE on OS X and MADV_DONTNEED on Linux to remove a physical page.
+  // https://bugs.chromium.org/p/chromium/issues/detail?id=823915
+#ifdef __APPLE__
+  const int flags = MADV_FREE_REUSABLE;
+#elif __linux__
+  const int flags = MADV_DONTNEED;
+#endif
+  auto ptr = get_page(page_id);
+  _unprotect_page(page_id);
+  if (madvise(ptr, page_id.byte_count(), flags) < 0) {
+    const auto error = errno;
+    Fail("Failed to call madvise(MADV_DONTNEED / MADV_FREE_REUSABLE): " + strerror(error));
+  }
+  _protect_page(page_id);
+  _madvice_free_call_count.fetch_add(1, std::memory_order_relaxed);
+}
+
+std::byte* VolatileRegion::get_page(const PageID page_id) {
+  DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+  const auto num_bytes = bytes_for_size_type(_size_type);
+  const auto data = _region.data() + page_id.index() * num_bytes;
+  return data;
+}
+
+Frame* VolatileRegion::get_frame(const PageID page_id) {
+  DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+  return &_frames[page_id.index()];
+}
+
+size_t VolatileRegion::memory_consumption() const {
+  return sizeof(*this) + sizeof(decltype(_frames)::value_type) * _frames.capacity();
+}
+
+size_t VolatileRegion::size() const {
+  return _frames.size();
+}
+
+PageSizeType VolatileRegion::size_type() const {
+  return _size_type;
+}
+
+void VolatileRegion::reuse(const PageID page_id) {
+  _unprotect_page(page_id);
+  // On OS X, we can use MADV_FREE_REUSE to update the memory accounting.
```

Comment:
I don't remember where I noted that already, but there is no OS X anymore. It's called "macOS".